ansible_base/lib/routers/association_resource_router.py

```diff
@@ -1,29 +1,85 @@
 from rest_framework import routers, serializers, status
 from rest_framework.decorators import action
+from rest_framework.permissions import SAFE_METHODS
 from rest_framework.response import Response
 
+from ansible_base.rbac.permission_registry import permission_registry
+from ansible_base.rbac.policies import check_content_obj_permission, visible_users
+
 logger = logging.getLogger('ansible_base.lib.routers.association_resource_router')
 
 
-class QuerySetMixinBase:
+class RelatedListMixin:
+    def check_parent_object_permissions(self, request, parent_obj):
+        # Associate and disassociate is a POST request, list is GET
+        # the normal process of get_object --> check_object_permissions
+        # will not check "change" permissions to the parent object on POST
+        # this method checks parent change permission, view permission should be handled by filter_queryset
+        if (request.method not in SAFE_METHODS) and 'ansible_base.rbac' in settings.INSTALLED_APPS and permission_registry.is_registered(parent_obj):
+            return check_content_obj_permission(request.user, parent_obj)
+        return True
+
+    def get_parent_object(self):
+        """Modeled mostly after DRF get_object, but for the parent model
+
+        Like for /api/v2/organizations/<pk>/cows/, this returns the organization
+        with the specified pk.
+        """
+        parent_view = self.parent_viewset()
+        parent_view.request = self.request
+        queryset = parent_view.filter_queryset(parent_view.get_queryset())
+        filter_kwargs = {'pk': self.kwargs['pk']}
+
+        parent_obj = get_object_or_404(queryset, **filter_kwargs)
+
+        # May raise a permission denied
+        self.check_parent_object_permissions(self.request, parent_obj)
+
+        return parent_obj
+
     def get_queryset(self):
-        parent_pk = self.kwargs['pk']
-        parent_model = self.parent_view_model
+        parent_instance = self.get_parent_object()
+        return getattr(parent_instance, self.association_fk).all()
+
+
+def basic_association_serializer_factory(qs):
+    class AssociationSerializer(serializers.Serializer):
+        instances = serializers.PrimaryKeyRelatedField(queryset=qs, many=True)
 
-        try:
-            parent_instance = parent_model.objects.get(pk=parent_pk)
-        except parent_model.DoesNotExist:
-            return parent_model.objects.none()
+    return AssociationSerializer
 
-        child_queryset = getattr(parent_instance, self.association_fk).all()
-        return child_queryset
 
+def filtered_association_serializer_factory(cls, qs):
+
+    class AssociationSerializer(serializers.Serializer):
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args, **kwargs)
+            request = self.context['request']
+            self.fields['instances'] = serializers.PrimaryKeyRelatedField(
+                queryset=cls.access_qs(request.user, queryset=qs),
+                many=True,
```

Comment:
@slemrmartin you can here here I wound up using the method we discussed (for associating items you don't have view for) for the association endpoints generically.

Reply:
awesome